File: platform/ios/xcode/FullStacked/Instance.swift
//
//  Instance.swift
//  FullStacked
//
//  Created by Charles-Philippe Lepage on 2024-11-06.
//

import SwiftUI

class Instance {
    private let isEditor: Bool
    private let projectId: String
    
    
    init (projectId: String, isEditor: Bool = false) {
        self.isEditor = isEditor
        self.projectId = projectId
    }
    
    func callLib(payload: Data) -> Data {
        var data = Data()
        if(self.isEditor) {
            data.append(Data([1])) // isEditor
            data.append(0.toBytes()) // no project id
        } else {
            data.append(Data([0])) // not editor
            // TODO: append serialized project id
        }
        
        data.append(payload)
        
        var responsePtr = Data().ptr()
        let size = call(data.ptr(), Int32(data.count), &responsePtr)
        let responseDataPtr = UnsafeBufferPointer(start: responsePtr!.assumingMemoryBound(to: UInt8.self), count: Int(size))
        let responseData = Data(responseDataPtr)
        freePtr(responsePtr)
        
        return responseData
    }
}

File: platform/android/studio/app/src/main/java/org/fullstacked/editor/Instance.kt
package org.fullstacked.editor

import java.util.Arrays

class Instance(val projectId: String, val isEditor: Boolean = false) {
    private lateinit var headerRequest: ByteArray

    private external fun call(buffer: ByteArray): ByteArray

    init {
        if(this.isEditor) {
            this.headerRequest = byteArrayOf(
                1 // isEditor
            )
            this.headerRequest += numberToBytes(0) // no project id
        } else {
            this.headerRequest = byteArrayOf(
                0 // is not Editor
            )
            val idData = this.projectId.toByteArray()
            this.headerRequest += numberToBytes(idData.size)
            this.headerRequest += idData
        }
    }

    fun callLib(payload: ByteArray) : ByteArray {
        return this.call(this.headerRequest + payload)
    }
}